Created addCardWithExtraParams function

mgaruccio opened the pull request:
Was using your api wrapper in a project of mine (thanks btw!) and needed the ability to assign a member to the card at create time so I added that in to the .addcard method, also made sure to add handling for the member param not being passed to ensure backwards compatibility.  not sure if this is worth adding or if it's too niche a use case but figured I'd offer up the changes.

sinisterstuf commented:
Cool solution but wouldn't it be better to make a new function for this? Imagine if we handled more/all params this way, it basically doesn't scale beyond 1 param IMO.

mgaruccio commented:
Yea your probably right about that, possibly a second function that either directly accepts a query object, or that takes a generic object and maps it to a query object? that way it would be able to support any parameter that's needed.

sinisterstuf commented:
I actually just had a dumb wrapper function in mind but your idea sounds much more flexible. The `addCard` function could take an optional object as an argument and merge it into the query, maybe with [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)? That would give you the power to add/override params without having to edit the function.

mgaruccio commented:
Ok I updated my pull request to support passing an object as the first param, and then accept a callback as the second param.  sorry for the kind of messy git commit logs, had a few typo's and not 100% confident using the rebase functionality(this is my first PR).

sinisterstuf commented:
> messy git commit logs

That's no problem, you can [squash them together](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) as if they were done in 1 commit.

> object as the first param

I think it's pretty cool, this way old code keeps working and new code can just use it.

Wouldn't it be good to add this as a new last parameter though? Mainly because the name `name` is misleading (because in the case of extra params, it is not a "name") and this way you can add name (a required param) and extra query params without having to write `{name: "foo", ...}` in the first parameter (because _every_ call will need a name). You might as well call it extraParams and have it after callback `..., callback, extraParams = null)`. Although I'm not sure if it's good to break the tradition of having callback as the last parameter.

So, if you don't think this is a good idea, it might be better to write a new function `addCardWithMember` or even `addCardWithExtraParams` which takes special params with meaningful names (other than name/description) and leaves the existing `addCard` function as it is.

What do you think?

> this is my first PR

That's great! Keep it up! 💛 👍

mgaruccio commented:
Had actually considered doing it that way but didn't want to break with the pattern of having callback be the last param.  Thinking about it some more I'm leaning towards a new `addCardWithExtraParams` function as most likely the best way to handle this, but happy to defer to your judgment on the simplest/cleanest way to implement.

sinisterstuf commented:
I think the query merging solution would be really cool, but it's usually better to go with a clear solution than a cool solution and I think here a new function would probably be clearer. Ultimately it's @norberteder whose judgement matters on this.

mgaruccio commented:
Ok, pushed a new update which removes the extra handling from addCard and created a new 'addCardWithExtraParams' function which I think accomplishes both my goal and also handles the request in  issue #27 any feedback is appreciated.

sinisterstuf commented:
Great! I think this implementation is good! ✅ 😃 

Some small suggestions:

- `Object.assign` modifies the first argument in-place, so there is no need for you to assign the result to `query`, `query` already has the new values
- would you like to add spaces after the commas in the argument list (this is a tidy detail but would keep formatting consistent with the rest of the code)
- also the indentation, you've got 2 spaces, everything else has 4 (again, a minor detail, but keeps things consistent)
- have you considered writing [some test cases](https://github.com/norberteder/trello/blob/master/test/spec.js#L143) for this new feature?

mgaruccio commented:
Thanks for the suggestions!   I cleaned up the code to fit correctly with the style of the rest of the library and added some basic test cases.  Let me know if there's anything else you think should be adjusted.

sinisterstuf commented:
No, I don't think so. The code is fine and I can see it runs perfectly on Node v4. We should update the test configuration to use newer versions of Node because Node versions v0.10 and v0.12 are end-of-life and no longer maintained since October and December last year respectively. Node v4 is currently the oldest LTS version so we should test it and the newer versions.

I'll get back to you on this. 😎

sinisterstuf commented:
@mgaruccio I updated the Travis config to run tests on Node v4 and v6. When #47 is merged you can update your branch to the master branch and the tests should run fine.

This would also be a good opportunity to squash your 12 commits into 1, since the pull request has deviated quite a lot from its initial intentions. Because the hashes change, you'll have to force the push to GitHub.

The following commands should do the trick:

```bash
git fetch origin master:master
git reset --hard origin/master
git merge --squash HEAD@{1}
git commit -m 'Write here some commit message for all 12 commits'
git push --force
```

Here I've assumed that you have no uncomitted changes (they would be lost) and your origin remote is norberteder/trello on github.com, if it isn't, use that one instead of "origin".

If anything goes wrong I have a local copy of your changes. 😉

sinisterstuf commented:
One last tiny thing, in the new tests you added, there's trailing whitespace after the semicolon [on line 208](https://github.com/norberteder/trello/pull/46/files#diff-8330e164d8515979491567746e4dc5c7R208), please delete it. Apart from that, it's perfect! 👍

mgaruccio commented:
Thank you for the git help!  Very much appreciated.  Requested updates have been made and pushed up to my repository.  let me know if you catch anything else you'd like me to update.

sinisterstuf commented:
You're welcome! Just keep an eye on #47, when it's merged to master you can rebase your branch to `origin/master` to make the tests green. 💚
